hive-metastore/src/test/java/org/apache/iceberg/hive/TestClientPool.java

```diff
@@ -51,4 +81,918 @@ protected void close(Object client) {
 
     }
   }
+
+  private static class MockHiveClientPool extends ClientPool<IMetaStoreClient, Exception> {
+
+    MockHiveClientPool(int poolSize, Class<? extends Exception> reconnectExc) {
+      super(poolSize, reconnectExc);
+    }
+
+    @Override
+    protected IMetaStoreClient newClient() {
+      return new MockMetaStoreFailureClient();
```

Comment:
Instead of creating our own client we might want to use Mockito to generate an object for us. Some good examles can be found here: https://github.com/apache/iceberg/blob/915a17573f2819992c866c9a76fbbf9612ca1d4a/aws/src/test/java/org/apache/iceberg/aws/glue/GlueCatalogTest.java

We can say that if a given method is called then return the given value. We can even define that the 2nd call should return another value.
Also when checking the results we can check the number of calls for the given methods.

Thanks for creating the test case!

Reply:
I tried it with `Mockito`, like this:
```
HiveClientPool clients = Mockito.mock(HiveClientPool.class);
Mockito.doReturn(Lists.newArrayList("t1", "t2")).when(clients)
  .run(client -> client.getAllTables(Mockito.any(String.class)));

hiveCatalog.listTables(Namespace.of("db1"))
```
However,  the mock result is `null`  when call  `.run(client -> client.getAllTables(Mockito.any(String.class)))`.
When  i use  `.run(Mockito.any(ClientPool.Action.class))` instead of `.run(client -> client.getAllTables(Mockito.any(String.class)))` , the reult is OK.

So, How should i test the different client APIs @pvary ?

Reply:
@Jecarm: I would do something like this:
```
    @Override
    protected IMetaStoreClient newClient() {
      IMetaStoreClient client = Mockito.mock(IMetaStoreClient.class);
      try {
        Mockito.doThrow(TTransportException.class).when(client).getAllDatabases();
        Mockito.doThrow(new MetaException("Got exception: org.apache.thrift.transport.TTransportException")).when(client).getAllFunctions();
        Mockito.doThrow(new MetaException("Another meta exception")).when(client).getTables(Mockito.anyString(), Mockito.anyString());
      } catch (Exception e) {
        throw new RuntimeException("This should not happen", e);
      }

      return client;
    }

    @Override
    protected IMetaStoreClient reconnect(IMetaStoreClient client) {
      IMetaStoreClient newClient = Mockito.mock(IMetaStoreClient.class);
      try {
        Mockito.doReturn(new ArrayList<>()).when(newClient).getAllDatabases();
        Mockito.doReturn(new GetAllFunctionsResponse()).when(newClient).getAllFunctions();
        Mockito.doReturn(new ArrayList<>()).when(newClient).getAllTables(Mockito.anyString());
      } catch (Exception e) {
        throw new RuntimeException("RE", e);
      }

      return newClient;
    }
```

After more thought I think checking the number of reconnection calls through the number of calls on the mock objects could be misleading, so I think we could check them in the `MockClientPool` in a counter / boolean if we do not find a more elegant way.

Reply:
Maybe something like this:
```
  private static class MockHiveClientPool extends HiveClientPool {
    public int reconnectCalls = 0;

    MockHiveClientPool() {
      super(2, new HiveConf());
    }

    @Override
    protected HiveMetaStoreClient newClient() {
[...]
    }

    @Override
    protected HiveMetaStoreClient reconnect(HiveMetaStoreClient client) {
      HiveMetaStoreClient client2 = Mockito.mock(HiveMetaStoreClient.class);
[..]
    }
  }
```

Reply:
Even better, we do not have to create a new class for the tests, but we can use `Mockito.spy()` on `HiveClientPool` and return the different clients on `newClient` and `reconnect`

Reply:
I refactored the test unit, please review it.

Reply:
Looks good. Thanks!